src/tpm2/NV_spt.c

```diff
@@ -216,3 +231,251 @@ TPM2B_NAME* NvGetIndexName(
     return name;
 }
 
+//*** NvPublic2FromNvPublic()
+// This function converts a legacy-form NV public (TPMS_NV_PUBLIC) into the
+// generalized TPMT_NV_PUBLIC_2 tagged-union representation.
+TPM_RC NvPublic2FromNvPublic(
+			     TPMS_NV_PUBLIC*   nvPublic,  // IN: the source S-form NV public area
+			     TPMT_NV_PUBLIC_2* nvPublic2  // OUT: the T-form NV public area to populate
+			     )
+{
+    TPM_HT handleType = HandleGetType(nvPublic->nvIndex);
+
+    switch(handleType)
+	{
+	  case TPM_HT_NV_INDEX:
+	    nvPublic2->nvPublic2.nvIndex = *nvPublic;
+	    break;
+	  case TPM_HT_PERMANENT_NV:
+	    nvPublic2->nvPublic2.permanentNV = *nvPublic;
+	    break;
+#if EXTERNAL_NV
+	  case TPM_HT_EXTERNAL_NV:
+	      {
+		  TPMS_NV_PUBLIC_EXP_ATTR* pub = &nvPublic2->nvPublic2.externalNV;
+
+		  pub->attributes = ExpandedAttributesFromLegacy(nvPublic->attributes);
+		  pub->authPolicy = nvPublic->authPolicy;
+		  pub->dataSize   = nvPublic->dataSize;
+		  pub->nameAlg    = nvPublic->nameAlg;
+		  pub->nvIndex    = nvPublic->nvIndex;
+		  break;
+	      }
+#endif
+	  default:
+	    return TPM_RCS_HANDLE;
+	}
+
+    nvPublic2->handleType = handleType;
+    return TPM_RC_SUCCESS;
+}
+
+//*** NvPublicFromNvPublic2()
+// This function converts a tagged-union NV public (TPMT_NV_PUBLIC_2) into the
+// legacy TPMS_NV_PUBLIC representation. This is a lossy conversion: any
+// bits in the extended area of the attributes are lost, and the Name cannot be
+// computed based on it.
+TPM_RC NvPublicFromNvPublic2(
+			     TPMT_NV_PUBLIC_2* nvPublic2,  // IN: the source T-form NV public area
+			     TPMS_NV_PUBLIC*   nvPublic    // OUT: the S-form NV public area to populate
+			     )
+{
+    switch(nvPublic2->handleType)
+	{
+	  case TPM_HT_NV_INDEX:
+	    *nvPublic = nvPublic2->nvPublic2.nvIndex;
+	    break;
+	  case TPM_HT_PERMANENT_NV:
+	    *nvPublic = nvPublic2->nvPublic2.permanentNV;
+	    break;
+#if EXTERNAL_NV
+	  case TPM_HT_EXTERNAL_NV:
+	      {
+		  TPMS_NV_PUBLIC_EXP_ATTR* pub = &nvPublic2->nvPublic2.externalNV;
+
+		  nvPublic->attributes = LegacyAttributesFromExpanded(pub->attributes);
+		  nvPublic->authPolicy = pub->authPolicy;
+		  nvPublic->dataSize   = pub->dataSize;
+		  nvPublic->nameAlg    = pub->nameAlg;
+		  break;
+	      }
+#endif
+	  default:
+	    return TPM_RCS_HANDLE;
+	}
+
+    return TPM_RC_SUCCESS;
+}
+
+//*** NvDefineSpace()
+// This function combines the common functionality of TPM2_NV_DefineSpace and
+// TPM2_NV_DefineSpace2.
+TPM_RC NvDefineSpace(TPMI_RH_PROVISION authHandle,
```

Comment:
Upstream seem incomplete since existing TPM_Define_Space should also use this function.